src/Type/InspectorTypeExtension.php

```diff
@@ -107,10 +107,23 @@ private function specifyAssertAll(MethodReflection $staticMethodReflection, Stat
             return new SpecifiedTypes();
         }
 
+        // In a negation context, we cannot precisely narrow types because we do
+        // not know the exact logic of the callable function. This means we
+        // cannot safely return 'mixed~iterable' since the value might still be
+        // a valid iterable.
+        //
+        // For example, a negation context with an 'is_string(...)' callable
+        // does not necessarily mean that the value cannot be an
+        // 'iterable<int>'. In such cases, it is safer to skip type narrowing
+        // altogether to prevent introducing new bugs into the code.
+        if ($context->false()) {
+            return new SpecifiedTypes();
+        }
+
         return $this->typeSpecifier->create(
             $node->getArgs()[1]->value,
             new IterableType(new MixedType(), new MixedType()),
-            TypeSpecifierContext::createTruthy(),
+            $context,
             false,
             $scope,
         );
```

Comment:
@mglaman now it works much safer:

```
// Inspector::assertAll()
$callable = fn (string $value): bool => $value === 'foo';
$input = mixed_function();
if (Inspector::assertAll($callable, $input)) {
    assertType("iterable", $input);
}
else {
    assertType('mixed', $input);
}
```

We can only confirm that it is `iterable` if the result is `TRUE`, but when it's `FALSE`, we cannot be sure what the type is, so we should keep its original type, which is `mixed`. 

It might be possible to narrow it based on the callable, but I think for now it is too complex and not worth the effort. We can always improve it later.

Reply:
This works for me, and good call on the comment